mindmeld/augmentation.py

```diff
@@ -241,6 +262,167 @@ def __init__(self, batch_size, language, paths, path_suffix, resource_loader):
+        running_start = 0
+        previous_end = 0
+        replaced_spans_entities = []
+        for (entity, span) in entity_matches:
+            if not running_start:
+                # First entity in the query
+                running_start = span.start
+                # Append text seen so far
+                new_paraphrase_text.append(paraphrase_text[:running_start])
+            else:
+                # Calculate new start based on previously replaced entity length
+                running_start += (span.start - previous_end)
+                # Append text seen between entities
+                new_paraphrase_text.append(paraphrase_text[previous_end:span.start])
+            gaz = None
+            # If not a system entity and gazetteer is available, load it
+            if not Entity.is_system_entity(entity.type):
+                gaz = self._resource_loader.get_gazetteer(entity.type)['entities']
+            if gaz:
+                # Create new Entity and Span based on random gaz entry for entity type
+                random_entity = random.sample(gaz, 1)[0]
+                new_span = Span(start=running_start, end=running_start + len(random_entity) - 1)
+                new_entity = Entity(text=random_entity,
+                                    entity_type=entity.type,
+                                    role=entity.role,
+                                    value=None)
+            else:
+                new_entity = entity
+                new_span = Span(start=running_start, end=running_start + len(entity.text) - 1)
+            running_start += len(new_entity.text)
+            replaced_spans_entities.append([new_entity, new_span])
+            new_paraphrase_text.append(new_entity.text)
+            previous_end = span.end + 1
+        # Append any leftover text
+        new_paraphrase_text.append(paraphrase_text[previous_end:])
+
+        processed_query = self._resource_loader.query_factory.create_query(
+                                                                  ''.join(new_paraphrase_text))
+        final_entities = []
+        for (entity, span) in replaced_spans_entities:
+            query_entity = QueryEntity.from_query(query=processed_query, span=span, entity=entity)
+            final_entities.append(query_entity)
+        return ProcessedQuery(query=processed_query, entities=tuple(final_entities))
+
+    def _annotate_entities(self, paraphrases, processed_queries):
+        """Annotates entities in the generated paraphrases with the entities in the original query
+
+            Args:
+                paraphrases (list(str)): List of unannotated paraphrases of queries
+                processed_queries (list(ProcessedQuery)): List of their corresponding original ProcessedQuery
+
+            Return:
+                paraphrases (list(str)): List of paraphrased queries.
+        """
+        valid_paraphrases = []
+        for i in range(len(processed_queries)):
```

Comment:
Nit: Alternatively consider

`for i, processed_query in enumerate(processed_queries):`

And changing `processed_queries[i]` to `processed_query` in line 388